lib/src/workspace_store.rs

```diff
@@ -0,0 +1,231 @@
+                .map_err(SimpleWorkspaceStoreError::Path)?;
+        }
+
+        Ok(store)
+    }
+
+    fn lock(&self) -> Result<FileLock, SimpleWorkspaceStoreError> {
+        Ok(FileLock::lock(self.lock_file.clone())?)
+    }
+
+    fn read_store(&self) -> Result<workspace_store::Workspaces, SimpleWorkspaceStoreError> {
+        let workspace_data = fs::read(&self.store_file).context(&self.store_file)?;
+
+        let workspaces_proto = workspace_store::Workspaces::decode(&*workspace_data)?;
+
+        Ok(workspaces_proto)
+    }
+
+    fn write_store(
+        &self,
+        workspaces_proto: workspace_store::Workspaces,
+    ) -> Result<(), SimpleWorkspaceStoreError> {
+        // We had created the store dir in load(), so parent() must exist.
+        let store_file_parent = self.store_file.parent().unwrap();
+        let temp_file = NamedTempFile::new_in(store_file_parent).context(store_file_parent)?;
+
+        temp_file
+            .as_file()
+            .write_all(&workspaces_proto.encode_to_vec())
+            .context(temp_file.path())?;
+
+        persist_temp_file(temp_file, &self.store_file).context(&self.store_file)?;
+
+        Ok(())
+    }
+}
+
+impl WorkspaceStore for SimpleWorkspaceStore {
+    fn name(&self) -> &'static str {
+        "simple"
+    }
+
+    fn add(&self, workspace_name: &WorkspaceName, path: &Path) -> Result<(), WorkspaceStoreError> {
+        let _lock = self.lock()?;
+
+        let mut workspaces_proto = self.read_store()?;
+
+        // Delete any existing entry with the same name
+        workspaces_proto
+            .workspaces
+            .retain(|w| w.name.as_str() != workspace_name.as_str());
+
+        workspaces_proto
+            .workspaces
+            .push(workspace_store::Workspace {
+                name: workspace_name.as_str().to_string(),
+                path: path_to_bytes(path)
+                    .map_err(SimpleWorkspaceStoreError::BadPathEncoding)?
+                    .into(),
+            });
+
+        self.write_store(workspaces_proto)?;
+
+        Ok(())
+    }
+
+    fn forget(&self, workspace_names: &[&WorkspaceName]) -> Result<(), WorkspaceStoreError> {
+        let _lock = self.lock()?;
+
+        let mut workspaces_proto = self.read_store()?;
+
+        workspaces_proto.workspaces.retain(|w| {
+            !workspace_names
+                .iter()
+                .any(|name| w.name.as_str() == name.as_str())
+        });
+
+        self.write_store(workspaces_proto)?;
+
+        Ok(())
+    }
+
+    fn rename(
+        &self,
+        old_name: &WorkspaceName,
+        new_name: &WorkspaceName,
+    ) -> Result<(), WorkspaceStoreError> {
+        let _lock = self.lock()?;
+
+        let mut workspaces_proto = self.read_store()?;
+
+        for workspace in &mut workspaces_proto.workspaces {
+            if workspace.name.as_str() == old_name.as_str() {
+                workspace.name = new_name.as_str().to_string();
+            }
+        }
```

Comment:
If we don't find `old_name`, it seems better to let the caller know by returning a `WorkspaceStoreError::NotFound` or similar